### PR TITLE
ci: use a wildcard subdomain for PR deployments

### PR DESCRIPTION
**Pull request by matifali**
Wildcard apps now work as expected.
image
We were hitting letsencrypt rate limits for certs because of deleting and recreating certs with each deployment. I tried to fix that by creating certificates in a shared namespace and copying them to the PR namespace.

### CODE REVIEW

File: .github/workflows/pr-cleanup.yaml [hunk 1]
```diff
@@ -48,3 +48,20 @@ jobs:
       - name: "Remove PR namespace"
         run: |
           kubectl delete namespace "pr${{ steps.pr_number.outputs.PR_NUMBER }}" || echo "namespace not found"
+
+      - name: "Remove DNS records"
+        run: |
+          set -euxo pipefail
+          # Get identifier for the record
+          record_id=$(curl -X GET "https://api.cloudflare.com/client/v4/zones/${{ secrets.PR_DEPLOYMENTS_ZONE_ID }}/dns_records?name=*.pr${{ steps.pr_number.outputs.PR_NUMBER }}.${{ secrets.PR_DEPLOYMENTS_DOMAIN }}" \
```
Comment (johnstcn) [inline .github/workflows/pr-cleanup.yaml:56]: nit: suggest URL-encoding the asterisk (`%2A`) in the `name` parameter

File: .github/workflows/pr-deploy.yaml [hunk 2]
```diff
@@ -168,35 +184,32 @@ jobs:
...
+                kind: ClusterIssuer
+              dnsNames:
+              - "${{ env.PR_DEPLOYMENT_ACCESS_URL }}"
+              - "*.${{ env.PR_DEPLOYMENT_ACCESS_URL }}"
           EOF
-          kubectl apply -f ingress.yaml
+          else
+            echo "Certificate exists. Skipping certificate creation."
+            echo "Copy certificate from pr-deployment-certs to pr${{ env.PR_NUMBER }} namespace"
+            kubectl get secret pr${{ env.PR_NUMBER }}-tls -n pr-deployment-certs -o yaml | sed 's/pr-deployment-certs/pr${{ env.PR_NUMBER }}/g' | kubectl apply -f -
```
Comment (johnstcn) [inline .github/workflows/pr-deploy.yaml:211]: Generally I find you need to delete some metadata fields from the resulting object (for example, `.metadata.uid`, `.metadata.resourceVersion`, `.metadata.creationTimestamp`). This is probably simpler to do with `jq`.

File: .github/workflows/pr-deploy.yaml [hunk 3]
```diff
@@ -294,16 +328,16 @@ jobs:
           echo "::add-mask::$password"
           echo "password=$password" >> $GITHUB_OUTPUT
 
-          /home/runner/coder login \
-            --first-user-username pr${{ env.PR_NUMBER }} \
-            --first-user-email ${{ env.PR_NUMBER }}@coder.com \
+          coder login \
```
Comment (johnstcn) [inline .github/workflows/pr-deploy.yaml:331]: nit: use `--global-config` to store the resulting tokens etc. in a well-known location

Comment (matifali) [reply .github/workflows/pr-deploy.yaml:331]: Doesn't it defaults to `.config/coderv2`? Or we should use another?

Comment (johnstcn) [reply .github/workflows/pr-deploy.yaml:331]: Yes, to `${HOME}/.config/coderv2` - my only thought is to have some method of separating the credentials for each PR env.

Comment (matifali) [reply .github/workflows/pr-deploy.yaml:331]: This is fine as it is run in GitHUb actions runner, and these are isolated and ephemeral for each run.

File: .github/workflows/pr-deploy.yaml (hunk 2 above)
Comment (matifali) [reply .github/workflows/pr-deploy.yaml:211]: or I found a better way. Thanks for pointing it out.

File: .github/workflows/pr-cleanup.yaml (hunk 1 above)
Comment (matifali) [reply .github/workflows/pr-cleanup.yaml:56]: Thank you :)

**Review by johnstcn — APPROVED** [review]
This looks fine to me, but we will need to be careful of LetsEncrypt rate limits. If we run into further issues we may need to look into using self-signed certificates.

**Comment by matifali** [comment]
The letsencrypt is 50 certs per domain per week. Now we have a different domain for each PR and are preserving the certificates by creating them in a shared namespace to avoid recreating them if one already exists.

**Pull request merged**